Add blog post on moving private projects to open source (#454)
ombulabs/blog: _posts/2020-09-15-open-sourcing-a-private-project.markdown

To completely delete the history including Pull Requests it is necessary to create a fresh repository. If you didn’t have any sensitive information to hide, and there is no problem with sharing the commit history, you can simply go to settings on the repository page, enter the danger zone at the bottom of settings and click on `Change repository visibility`.

If you do need to hide the history of the project there are several ways to do this, but this is the way we used:

Note: before beginning this we made a copy on GitHub and locally in case anything went wrong.
```
$ git pull origin <main branch for repository >
```
Follow steps on [GitHub](https://docs.github.com/en/enterprise/2.17/user/github/administering-a-repository/deleting-a-repository) to delete the repository.

Follow steps on [GitHub](https://docs.github.com/en/github/getting-started-with-github/create-a-repo) to create a new repository and make sure to choose the public option when choosing the visibility, and don’t include the README.
```

> # Choose which commit you would like to include in the new tree(we chose the first commit. Use git log and copy the commit hash.

Reviewer: I'd say that if you are looking to squash the last number of commits into a single one in the commit history, we can use the `git reset --soft [commit]` option. In our case we squashed all commits against the first one, but deleting the repo and recreating it would have given the same effect. There's an interesting side effect in this approach though, Github seems to detect contributions from other developers while if it was a repo from scratch it wouldn't.

$ git reset --soft <commit hash>
$ git commit --amend
$ adjust commit title as needed

> $ git remote add origin <clone new GitHub repository>

Reviewer: This step was done slightly differently. We renamed the old remote and then we added the new one with (found these lines in my bash history, thanks bash!):

```
git remote rename origin oldrepo
git remote add origin git@github.com:fastruby/points.git
```

---
layout: post

> title: "How to Open Source a Closed Source Project"

Reviewer 2: "Private Project" sounds more common to me than "Closed Source Project". But either one looks good 👍

> date: 2020-09-17 11:00:00

Reviewer 3: @FionaDL This will need to be updated when you're done addressing feedback. 👍

> categories: ["open source"]

Author: ```suggestion
categories: ["open-source"]
```

> author: "fionaDL"

Reviewer 2: This should be:

```suggestion
author: "fionadl"
```

Reviewer 2: @FionaDL btw, if you prefer to display `fionaDL`, you can update it in the config and in the older posts

Author: @lubc It doesn't matter to me :)

---

Open Source is one of the things that we [value]( https://www.ombulabs.com/blog/values/our-values.html) as a company. Our philosophy is that “everything we do should be open by default”. This is why in the past few months we decided to open source two of our internal projects. We built [Dash](https://github.com/fastruby/dash) and [Points](https://github.com/fastruby/points) to make our daily processes smoother, and then we thought “hey, why not share them with everyone?”.
Now we also want to share the process of how we turned them from private to open source.

<!--more-->

> ## Making a list

Author: ```suggestion
### Making a list
```

> ### Making a list

Reviewer 3: @FionaDL Could you make sure you use h2 instead of h3? In terms of readability and SEO, it makes more sense to follow always the same hierarchy in our articles. The title is h1. 

```suggestion
## Making a list
```

Before we began the process of open sourcing we decided to make a list. We thought about what it would take to move the projects into the eyes of the public. We thought about what we wanted the app to look like, if there were any “secrets” in the code, and how we would present the projects. These are some of the things we came up with for our checklist.

### Hiding sensitive information

> It is important to go through commit history and the code to make sure there are no “secrets” you would be sharing with the public when you open source. We ended up needing to squash the commit history because we had some areas that had sensitive information. Maybe your code is already free of sensitive information, but it is a good idea to check.

Reviewer 3: ```suggestion
It is important to go through commit history and the code to make sure there are no “secrets” you would be sharing with the public when you publish your source code and Git history. We ended up needing to squash the commit history because we had some areas that had sensitive information. Maybe your code is already free of sensitive information, but it is a good idea to check.
```

> Look at places where you may have hardcoded company specific or personal information. We often move these hardcoded pieces of code into environment variables. Just make sure you include the environment variables and an explanation in the README.

Reviewer 3: This sounds a bit confusing: Just make sure you include the environment variables and an explanation in the README.

I think you mean something like this: Just make sure you include a sample file with the environment variables (and sample values) and an explanation in the README. Something like `.env.sample` so future users and contributors have an idea of what those look like.

### Creating a comprehensive README

Having a good README is fundamental when creating an open source project. It is important for it to be robust and filled with easy to understand information on setting up and using the project. Some things that are important to include:

> - Description of the Project

Reviewer 3: This needs to be changed if we want to be consistent. In the line below you write "project" with lowercase P. 

```suggestion
- Description of the project
```

- Steps on how to setup the project
- A contribution guide
- Code of Conduct
- Sponsorship (if applicable)
- License

> You can read more about this over [here](https://www.fastruby.io/blog/open-source/ombulabs-open-source-guidelines.html) in our article about open source practices.

Reviewer 3: I think this sounds a little better and it is more SEO-friendly 😄 

```suggestion
You can read more about this in [our article about open source practices](https://www.fastruby.io/blog/open-source/ombulabs-open-source-guidelines.html).
```

The next step in open sourcing the project is to move it from Private to Public. We use [GitHub](https://github.com/) to store our repositories, so these steps are the ones for GitHub.
To completely delete the history including Pull Requests it is necessary to create a fresh repository. If you didn’t have any sensitive information to hide, and there is no problem with sharing the commit history, you can simply go to settings on the repository page, enter the danger zone at the bottom of settings and click on `Change repository visibility`.

If you do need to hide the history of the project there are several ways to do this, but this is the way we used:

Note: before beginning this we made a copy on GitHub and locally in case anything went wrong.
```
$ git pull origin <main branch for repository >
```
Follow steps on [GitHub](https://docs.github.com/en/enterprise/2.17/user/github/administering-a-repository/deleting-a-repository) to delete the repository.

Follow steps on [GitHub](https://docs.github.com/en/github/getting-started-with-github/create-a-repo) to create a new repository and make sure to choose the public option when choosing the visibility, and don’t include the README.

> ```

Reviewer 2: @FionaDL Styles are breaking up a bit here. You might need to add an empty line before this.

![Screen Shot 2020-10-05 at 3 23 15 PM](https://user-images.githubusercontent.com/7442887/95117289-ba3e7200-071e-11eb-9a62-e4c26151fdd1.png)

# Choose which commit you would like to include in the new tree(we chose the first commit). Use git log and copy the commit hash.
$ git reset --soft <commit hash>
$ git commit --amend
# adjust commit title as needed
$ git remote rename origin <oldrepo>
$ git remote add origin <clone new GitHub repository>
$ git push origin <main branch>
```
If you are looking to squash the last number of commits into a single one in the commit history, we can use the git reset --soft [commit] option. In our case we squashed all commits against the first one, but deleting the repo and recreating it would have given the same effect. There's an interesting side effect in this approach though, GitHub seems to detect contributions from other developers while if it was a repo from scratch it wouldn't.

### Creating a GitHub Page

> A GitHub page is not essential when open sourcing a project, but it is a good idea. We have GitHub Pages for several of our projects including; [Dash](https://fastruby.github.io/dash/), [Points](https://fastruby.github.io/points/), and [Pecas](https://fastruby.github.io/pecas/).

Reviewer 3: @FionaDL Why is it a good idea? Maybe you can explain that some people like to see screenshots of the application to better understand what it does and what it looks like?

Steps we used to create a GitHub Page

> ```

Reviewer 2: Same here

![Screen Shot 2020-10-05 at 3 24 58 PM](https://user-images.githubusercontent.com/7442887/95117436-fd004a00-071e-11eb-9835-6ed78956f0e4.png)

$ git checkout --orphan gh-pages
# preview files to be deleted
$ git rm -rf --dry-run .
# actually delete the files
$ git rm -rf .
# create an index.html file
# add any styling you want to associate for the file
# Push gh-pages branch to GitHub
```
Go into settings for the repository and activate a Gh-Page. It will automatically use your gh-pages branch.

Note. This is for custom GitHub pages. It is also possible to quickly build a GitHub page using a theme and the already existing README.

### Badges
Adding Badges to the README is a great way to show pieces of information to potential users quickly and aesthetically. [Circle CI](https://circleci.com/docs/2.0/status-badges/), [Travis CI](https://docs.travis-ci.com/user/status-images/), [Code Climate](https://codeclimate.com/github/codeclimate/codeclimate/badges/) and more have badges or status images that you can embed in the README.

### Adding known issues to GitHub

> We tend to use [Pivotal Tracker](https://www.pivotaltracker.com/) for our internal projects to keep track of stories and known issues. Once we open source we like to close the Pivotal Tracker project and move all known issues or desired features to GitHub Issues. This way it will be a public history of everything being worked on in the project.

Reviewer 3: @FionaDL Not sure about this one. But @arielj and @rdormer have been working on a `standardrb` implementation with githooks that might be interesting for this project. Do we want to recommend implementing this for OSS projects? 

Also, to have issues and PR templates might be useful as well. I think this is something that @ABizzinotto worked on. Not sure if that info belongs in this article or in this one: https://www.fastruby.io/blog/open-source/ombulabs-open-source-guidelines.html

Author: @etagwerker after talking with @rdormer and @arielj we have decided to leave this out for now. I think we could discuss adding the templates to the other article. Luis and I worked on a template last week for points.

> We tend to use [Pivotal Tracker](https://www.pivotaltracker.com/) for our internal projects to keep track of stories and known issues. Once we open source we like to close the Pivotal Tracker project and move all known issues or desired features to GitHub Issues. This way it will be a public history of everything being worked on in the project.

Reviewer 2: @FionaDL it would be nice to add a link to https://www.ombulabs.com/blog/agile/pivotal-tracker/how-we-use-pivotal-tracker-at-ombu-labs.html when you mention Pivotal Tracker

### Announcing your new open source project

> At OmbuLabs we like to announce our projects with a blog post. We will also use Tweets, posts on LinkedIn or share anywhere else we feel is appropriate.

Reviewer 3: @FionaDL Maybe add a link to an example of this here?

Follow steps on [GitHub](https://docs.github.com/en/github/getting-started-with-github/create-a-repo) to create a new repository and make sure to choose the public option when choosing the visibility, and don’t include the README.
```
# Choose which commit you would like to include in the new tree(we chose the first commit). Use git log and copy the commit hash.
$ git reset --soft <commit hash>
$ git commit --amend
# adjust commit title as needed
$ git remote rename origin <oldrepo>
$ git remote add origin <clone new GitHub repository>
$ git push origin <main branch>
```

> If you are looking to squash the last number of commits into a single one in the commit history, we can use the git reset --soft [commit] option. In our case we squashed all commits against the first one, but deleting the repo and recreating it would have given the same effect. There's an interesting side effect in this approach though, GitHub seems to detect contributions from other developers while if it was a repo from scratch it wouldn't.

Reviewer: Looks good. 

I was thinking that not all contributors will appear (the ones in the intermediate commits), only the one that did the commit over which we squashed, but maybe its not even worth mentioning it.
